src/hotspot/cpu/x86/stubGenerator_x86_64_poly_mont.cpp

```diff
@@ -213,26 +219,277 @@ void montgomeryMultiply(const Register aLimbs, const Register bLimbs, const Regi
+  __ movq(modulus, mask52);
+  __ vmovdqu(Modulus, ExternalAddress(modulus_p256(1)), Assembler::AVX_256bit, rscratch);
+
+  // A = load(*aLimbs);
+  __ movq(a, Address(aLimbs, 0));
+  __ vmovdqu(A, Address(aLimbs, 8)); //Assembler::AVX_256bit
+
+  // Acc1 = 0
+  __ vpxorq(Acc1, Acc1, Acc1, Assembler::AVX_256bit);
+  for (int i = 0; i< 5; i++) {
+      // Acc2 = 0
+      __ vpxorq(Acc2, Acc2, Acc2, Assembler::AVX_256bit);
+
+      // B = replicate(bLimbs[i])
+      __ movq(tmp_rax, Address(bLimbs, i*8)); //(b==rax)
+      __ vpbroadcastq(B, Address(bLimbs, i*8), Assembler::AVX_256bit);
+
+      // Acc1 += A * B
+      // Acc2 += A *h B
+      __ mulq(a); // rdx:rax = a*rax
+      if (i == 0) {
+        __ movq(acc1, tmp_rax);
+        __ movq(acc2, tmp_rdx);
+      } else {
+        __ xorq(acc2, acc2);
+        __ addq(acc1, tmp_rax);
+        __ adcq(acc2, tmp_rdx);
+      }
+      __ vpmadd52luq(Acc1, A, B, Assembler::AVX_256bit);
+      __ vpmadd52huq(Acc2, A, B, Assembler::AVX_256bit);
+
+      // N = replicate(Acc1[0])
+      if  (i != 0) {
+        __ movq(tmp_rax, acc1); // (n==rax)
+      }
+      __ andq(tmp_rax, mask52);
+      __ movq(N, acc1); // masking implicit in evpmadd52
+      __ vpbroadcastq(N, N, Assembler::AVX_256bit);
+
+      // Acc1 += M *  N
+      __ mulq(modulus); // rdx:rax = modulus*rax
+      __ vpmadd52luq(Acc1, Modulus, N, Assembler::AVX_256bit);
+      __ addq(acc1, tmp_rax); //carry flag set!
+
+      // Acc2 += M *h N
+      __ adcq(acc2, tmp_rdx);
+      __ vpmadd52huq(Acc2, Modulus, N, Assembler::AVX_256bit);
+
+      // Combine high/low partial sums Acc1 + Acc2
+
+      // carry = Acc1[0] >> 52
+      __ shrq(acc1, 52); // low 52 of acc1 ignored, is zero, because Montgomery
+
+      // Acc2[0] += carry
+      __ shlq(acc2, 12);
+      __ addq(acc2, acc1);
+
+      // Acc1 = Acc1 shift one q element >>
+      __ movq(acc1, Acc1);
+      __ vpermq(Acc1, Acc1, 0b11111001, Assembler::AVX_256bit);
+      __ vpand(Acc1, Acc1, MaskLimb5, Assembler::AVX_256bit);
+
+      // Acc1 = Acc1 + Acc2
+      __ addq(acc1, acc2);
+      __ vpaddq(Acc1, Acc1, Acc2, Assembler::AVX_256bit);
+  }
+
+  __ movq(acc2, acc1);
+  __ subq(acc2, modulus);
+  __ vpsubq(Acc2, Acc1, Modulus, Assembler::AVX_256bit);
+  __ vmovdqu(Address(rsp, -32), Acc2); //Assembler::AVX_256bit
+
+  // Carry propagate the subtraction result first (since the last carry is used
+  // to select result)
+  // acc1  = tmp2;
+  // acc2  = tmp3;
+  // mask52 = tmp5
+  Register limb[] = {acc2, tmp1, tmp4, tmp_rdx, tmp6};
+  Register carry = tmp_rax;
+  for (int i = 0; i<5; i++) {
+    if (i > 0) {
+      __ movq(limb[i], Address(rsp, -32-8+i*8));
+      __ addq(limb[i], carry);
+    }
+    __ movq(carry, limb[i]);
+    if (i==4) break;
+    __ sarq(carry, 52);
+  }
+  __ sarq(carry, 63);
+  __ notq(carry); //select
+  Register select = carry;
+  carry = tmp7;
+
+  // Now carry propagate the multiply result and (constant-time) select correct
+  // output digit
```

Comment:
Carry propagate multiply result is done before subtracting modulus in the Java code.